include/sparrow/buffer_adaptor.hpp

```diff
@@ -0,0 +1,499 @@
+
+        constexpr const_iterator cbegin() const noexcept;
+        constexpr const_iterator cend() const noexcept;
+
+        constexpr reverse_iterator rbegin() noexcept;
+        constexpr reverse_iterator rend() noexcept;
+
+        constexpr const_reverse_iterator rbegin() const noexcept;
+        constexpr const_reverse_iterator rend() const noexcept;
+
+        constexpr const_reverse_iterator crbegin() const noexcept;
+        constexpr const_reverse_iterator crend() const noexcept;
+
+        // Capacity
+
+        [[nodiscard]] constexpr size_type size() const noexcept;
+        [[nodiscard]] constexpr size_type max_size() const noexcept;
+        [[nodiscard]] constexpr size_type capacity() const noexcept;
+        [[nodiscard]] constexpr bool empty() const noexcept;
+        constexpr void reserve(size_type new_cap);
+        constexpr void shrink_to_fit();
+
+        // Modifiers
+
+        constexpr void clear() noexcept;
+
+        constexpr iterator insert(const_iterator pos, const value_type& value);
+        constexpr iterator insert(const_iterator pos, size_type count, const value_type& value);
+        template <class InputIt>
+            requires std::input_iterator<InputIt>
+        constexpr iterator insert(const_iterator pos, InputIt first, InputIt last);
+        constexpr iterator insert(const_iterator pos, std::initializer_list<value_type> ilist);
+
+        template <class... Args>
+        constexpr iterator emplace(const_iterator pos, Args&&... args);
+
+        constexpr iterator erase(const_iterator pos);
+        constexpr iterator erase(const_iterator first, const_iterator last);
+
+        constexpr void push_back(const value_type& value);
+
+        constexpr void pop_back();
+
+        constexpr void resize(size_type new_size);
+        constexpr void resize(size_type new_size, const value_type& value);
+
+    private:
+
+        static constexpr double _sizeof_ratio = static_cast<double>(sizeof(buffer_reference_value_type))
+                                                / static_cast<double>(sizeof(value_type));
+
+        static constexpr auto _type_size = static_cast<size_type>(1. / _sizeof_ratio);
+
+        static constexpr size_type _extra_buffervalue_type_size = sizeof(buffer_reference_value_type)
+                                                                  % sizeof(value_type);
+
+        [[nodiscard]] constexpr buffer_reference::size_type index_for_buffer(size_type idx) const noexcept;
+
+        [[nodiscard]] buffer_reference::const_iterator get_buffer_reference_iterator(const_iterator pos);
+
+        buffer_reference& _buffer;
+    };
+
+    template <typename T, typename U>
+        requires(sizeof(U) < sizeof(T))
+    constexpr typename buffer_adaptor<T, U>::buffer_reference::size_type
+    buffer_adaptor<T, U>::index_for_buffer(size_type idx) const noexcept
+    {
+        return static_cast<buffer_reference::size_type>(static_cast<double>(idx) / _sizeof_ratio);
```

Comment:
- I think you can use `size_type` here.
- I think `idx * to_from_size_ratio` would be more efficient than a division.

Reply:
👍